src/clone.c

```diff
@@ -959,19 +983,36 @@ FuncDeclaration *buildDtor(AggregateDeclaration *ad, Scope *sc)
         ex = new DotVarExp(loc, ex, v, 0);
         if (v->type->toBasetype()->ty == Tstruct)
         {
-            // this.v.__dtor()
+            // this.v.__xdtor()
+
+            // This is a hack so we can call destructors on const/immutable objects.
+            ex = new AddrExp(loc, ex);
+            ex = new CastExp(loc, ex, v->type->mutableOf()->pointerTo());
+            ex = new PtrExp(loc, ex);
+            if (stc & STCsafe)
+                stc = (stc & ~STCsafe) | STCtrusted;
+
             ex = new DotVarExp(loc, ex, sdv->dtor, 0);
             ex = new CallExp(loc, ex);
         }
         else
         {
-            // Typeinfo.destroy(cast(void*)&this.v);
-            Expression *ea = new AddrExp(loc, ex);
-            ea = new CastExp(loc, ea, Type::tvoid->pointerTo());
-
-            Expression *et = new TypeidExp(loc, v->type);
-            et = new DotIdExp(loc, et, Id::destroy);
-            ex = new CallExp(loc, et, ea);
+            // _ArrayDtor((cast(S*)this.v.ptr)[0 .. n])
+
+            // This is a hack so we can call destructors on const/immutable objects.
+            ex = new DotIdExp(loc, ex, Id::ptr);
+            ex = new CastExp(loc, ex, sdv->type->pointerTo());
+            if (stc & STCsafe)
+                stc = (stc & ~STCsafe) | STCtrusted;
```

Comment:
Seems like this attribute change be visible on the type, e.g. `typeof(sd._xdtor)`?.
Maybe directly change `ex->type` instead of casting `.ptr` to avoid that